CDC panic: fetch events from an non-existent table

### What did you do?

Create kafka changefeed and run workload

### What did you expect to see?

no panic

### What did you see instead?


```
kubectl --kubeconfig kubeconfig.yml -n stable-testbed-47l4r logs -p maincluster-ticdc-1
[WARN] TiCDC server data-dir is not set. Please use `cdc server --data-dir` to start the cdc server if possible.
panic: fetch events from an non-existent table

goroutine 1323 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x3?, 0x3?, {0x0?, 0x0?, 0xc01bc043e0?})
        go.uber.org/zap@v1.25.0/zapcore/entry.go:196 +0x54
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc020e46340, {0xc01a3ff740, 0x3, 0x3})
        go.uber.org/zap@v1.25.0/zapcore/entry.go:262 +0x3ec
go.uber.org/zap.(*Logger).Panic(0xc010c71c00?, {0x4af71ed?, 0x0?}, {0xc01a3ff740, 0x3, 0x3})
        go.uber.org/zap@v1.25.0/logger.go:258 +0x51
github.com/pingcap/log.Panic({0x4af71ed?, 0xc0069649c0?}, {0xc01a3ff740?, 0x64c3?, 0x62bb?})
        github.com/pingcap/log@v1.1.1-0.20230317032135-a0d097d16e22/global.go:54 +0x85
github.com/pingcap/tiflow/cdc/processor/sourcemanager/engine/pebble.(*EventSorter).FetchByTable(0xc002daca50, {0x12c74, {0xc019d49d28, 0x12, 0x18}, {0xc019d49d40, 0x12, 0x18}}, {0x62c554ef870002f, 0x62c554ef870002f}, ...)
        github.com/pingcap/tiflow/cdc/processor/sourcemanager/engine/pebble/event_sorter.go:193 +0x4b9
github.com/pingcap/tiflow/cdc/processor/sourcemanager.(*SourceManager).FetchByTable(0xc0024e6000, {0x12c74, {0xc019d49d28, 0x12, 0x18}, {0xc019d49d40, 0x12, 0x18}}, {0x62c554ef870002f, 0x62c554ef870002f}, ...)
        github.com/pingcap/tiflow/cdc/processor/sourcemanager/manager.go:169 +0xb0
github.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask(0xc0069c2680, {0x55aa428, 0xc000e40280}, 0xc020aa6d20)
        github.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:214 +0x568
github.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTasks(0x24?, {0x55aa428, 0xc000e40280}, 0xc002a71080)
        github.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:104 +0xbd
github.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).startSinkWorkers.func1()
        github.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:400 +0x2a
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync@v0.3.0/errgroup/errgroup.go:75 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1318
        golang.org/x/sync@v0.3.0/errgroup/errgroup.go:72 +0x96

```

### Versions of the cluster

[root@maincluster-ticdc-0 /]# /cdc version
Release Version: v7.4.0
Git Commit Hash: 744211d931db53bdce429063a64f4db8bf4ad3bf
Git Branch: heads/refs/tags/v7.4.0
UTC Build Time: 2023-10-09 11:58:23
Go Version: go version go1.21.1 linux/amd64
Failpoint Build: false

--- a/cdc/processor/sinkmanager/table_sink_worker.go
+++ b/cdc/processor/sinkmanager/table_sink_worker.go
@@ -255,10 +255,7 @@ func (w *sinkWorker) handleTask(ctx context.Context, task *sinkTask) (finalErr e
 			allEventSize += size
 		}
 
-		if err := advancer.tryAdvanceAndAcquireMem(
-			false,
-			pos.Valid(),
-		); err != nil {
+		if err := advancer.tryAdvanceAndAcquireMem(false, pos.Valid()); err != nil {
 			return errors.Trace(err)
 		}
 	}

--- a/cdc/processor/sinkmanager/table_sink_wrapper.go
+++ b/cdc/processor/sinkmanager/table_sink_wrapper.go
@@ -62,10 +62,12 @@ type tableSinkWrapper struct {
 
 	// state used to control the lifecycle of the table.
 	state *tablepb.TableState
+
 	// startTs is the start ts of the table.
 	startTs model.Ts
 	// targetTs is the upper bound of the table sink.
 	targetTs model.Ts
+
 	// barrierTs is the barrier bound of the table sink.
 	barrierTs atomic.Uint64
 	// receivedSorterResolvedTs is the resolved ts received from the sorter.

--- a/cdc/processor/sourcemanager/engine/pebble/event_sorter.go
+++ b/cdc/processor/sourcemanager/engine/pebble/event_sorter.go
@@ -59,7 +59,6 @@ type EventSorter struct {
 // EventIter implements sorter.EventIterator.
 type EventIter struct {
 	tableID  model.TableID
-	state    *tableState
 	iter     *pebble.Iterator
 	headItem *model.PolymorphicEvent
 	serde    encoding.MsgPackGenSerde
@@ -146,6 +145,8 @@ func (s *EventSorter) RemoveTable(span tablepb.Span) {
 }
 
 // Add implements engine.SortEngine.
+//
+// Panics if the table doesn't exist.
 func (s *EventSorter) Add(span tablepb.Span, events ...*model.PolymorphicEvent) {
 	s.mu.RLock()
 	state, exists := s.tables.Get(span)
@@ -185,15 +186,18 @@ func (s *EventSorter) OnResolve(action func(tablepb.Span, model.Ts)) {
 
 // FetchByTable implements engine.SortEngine.
 func (s *EventSorter) FetchByTable(span tablepb.Span, lowerBound, upperBound engine.Position) engine.EventIterator {
+	iterReadDur := engine.SorterIterReadDuration()
+	eventIter := &EventIter{
+		tableID:      span.TableID,
+		serde:        s.serde,
+		nextDuration: iterReadDur.WithLabelValues(s.changefeedID.Namespace, s.changefeedID.ID, "next"),
+	}
+
 	s.mu.RLock()
 	state, exists := s.tables.Get(span)
 	s.mu.RUnlock()
-
 	if !exists {
-		log.Panic("fetch events from an non-existent table",
-			zap.String("namespace", s.changefeedID.Namespace),
-			zap.String("changefeed", s.changefeedID.ID),
-			zap.Stringer("span", &span))
+		return eventIter
 	}
 
 	sortedResolved := state.sortedResolved.Load()
@@ -208,21 +212,14 @@ func (s *EventSorter) FetchByTable(span tablepb.Span, lowerBound, upperBound eng
 	}
 
 	db := s.dbs[getDB(span, len(s.dbs))]
-	iterReadDur := engine.SorterIterReadDuration()
 
 	seekStart := time.Now()
 	iter := iterTable(db, state.uniqueID, span.TableID, lowerBound, upperBound)
 	iterReadDur.WithLabelValues(s.changefeedID.Namespace, s.changefeedID.ID, "first").
 		Observe(time.Since(seekStart).Seconds())
 
-	return &EventIter{
-		tableID: span.TableID,
-		state:   state,
-		iter:    iter,
-		serde:   s.serde,
-
-		nextDuration: iterReadDur.WithLabelValues(s.changefeedID.Namespace, s.changefeedID.ID, "next"),
-	}
+	eventIter.iter = iter
+	return eventIter
 }
 
 // FetchAllTables implements engine.SortEngine.
@@ -240,10 +237,7 @@ func (s *EventSorter) CleanByTable(span tablepb.Span, upperBound engine.Position
 	s.mu.RUnlock()
 
 	if !exists {
-		log.Panic("clean an non-existent table",
-			zap.String("namespace", s.changefeedID.Namespace),
-			zap.String("changefeed", s.changefeedID.ID),
-			zap.Stringer("span", &span))
+		return nil
 	}
 
 	return s.cleanTable(state, span, upperBound)
@@ -258,6 +252,8 @@ func (s *EventSorter) CleanAllTables(upperBound engine.Position) error {
 }
 
 // GetStatsByTable implements engine.SortEngine.
+//
+// Panics if the table doesn't exist.
 func (s *EventSorter) GetStatsByTable(span tablepb.Span) engine.TableStats {
 	s.mu.RLock()
 	state, exists := s.tables.Get(span)

--- a/cdc/processor/sourcemanager/engine/pebble/event_sorter_test.go
+++ b/cdc/processor/sourcemanager/engine/pebble/event_sorter_test.go
@@ -179,6 +179,6 @@ func TestCleanData(t *testing.T) {
 
 	span := spanz.TableIDToComparableSpan(1)
 	s.AddTable(span, 0)
-	require.Nil(t, s.CleanByTable(spanz.TableIDToComparableSpan(2), engine.Position{}))
+	require.NoError(t, s.CleanByTable(spanz.TableIDToComparableSpan(2), engine.Position{}))
 	require.Nil(t, s.CleanByTable(span, engine.Position{}))
 }

--- a/cdc/processor/sourcemanager/engine/pebble/event_sorter_test.go
+++ b/cdc/processor/sourcemanager/engine/pebble/event_sorter_test.go
@@ -179,8 +179,6 @@ func TestCleanData(t *testing.T) {
 
 	span := spanz.TableIDToComparableSpan(1)
 	s.AddTable(span, 0)
-	require.Panics(t, func() {
-		s.CleanByTable(spanz.TableIDToComparableSpan(2), engine.Position{})
-	})
+	require.Nil(t, s.CleanByTable(spanz.TableIDToComparableSpan(2), engine.Position{}))
 	require.Nil(t, s.CleanByTable(span, engine.Position{}))
 }

